doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,231 @@
+| `process_rate`       | [`uint8`] | The rate at which Materialize is processing upstream data.         |
+```
+
+This rate has units that depend on the source type (replication bytes for Postgres, offsets for Kafka, gtid's for MySQL) and are
+designed to be _compared_. The following charts show what these metrics will look like, depending on whether Materialize is keeping up
+or falling behind.
+
+![steady-state](./static/source_metrics_2/steady_state.png)
+
+
+Note that when our _process rate_ is consistently below the `upstream_rate`, the user knows their source isn't keeping up with upstream.
+
+### Implementation
+
+These rates are _source frontier differences_. Each source's `SourceReader` implementation will
+be required to expose a continuous stream of _upstream source frontiers_. This means:
+
+- The Kafka source will periodically fetch metadata and expose per-partition offset high-watermarks.
+- The MySQL source will periodically expose the result of `SELECT @gtid_executed`, which is the gtid of the latest transaction.
+- The Postgres source will periodically expose the result of `SELECT pg_current_wal_lsn()`, which is the lsn of the latest transaction.
+
+Additionally, the source pipeline will periodically invert the latest frontier we have committed for the source, from the Materialize timestamp
+domain to the source-timestamp domain. For MySQL and Postgres sources, this frontier will the `meet` of the subsource frontiers.
+
+`upstream_rate` will be derived from the two latest _upstream source frontiers_, by calculating _source frontier difference_ between them.
+Similarly `process_rate` will be derived from the two latest committed frontiers.
+
+
+### Existing metrics
+
+Existing metrics such as `messages_received` and `bytes_received` continue to be useful for understanding the general performance of a given source.
+If users find some of them confusing or not useful, we can remove them in the future.
+
+## Minimal Viable Prototype
+
+N/A for now. This design document is primarily designed to expose
+
+
+## Alternatives
+
+- Expose 100% accurate information about how much of the snapshot we have processed, as opposed to how much we have read.
+    - Rejected as it's considered as difficult as implementing Hydration Backpressure.
+- Expose progress information in a different format per source type.
+    - Rejected as it adds undue complexity in the system catalog, console, and for the user to interpret.
+        - In the future, if customers want per-partition information for Kafka sources, we can do so easily using this same general design.
+- Consolidate the two metric sets proposed in this design document into one.
+    - Rejected due to its complexity.
+- Expose more information to users about their sources.
+    - This design document is intentionally constrained, to help users answer 2 well-known questions about their sources.
+
```

Comment:
I think it's worth calling out here that (I think) https://github.com/MaterializeInc/materialize/issues/23345 is an alternative to (2), in that if we always reclocked to the latest upstream offset, you should be able to observe whether a source is keeping up with upstream by looking for skew between the progress collection and the data collection. The downside of this alternative is that it couples progress to a policy decision about reclocking; what's nice about what you proposed here is that progress tracking is totally independent of reclocking policy.

Reply:
Its an alternative in the sense that we would get the _upstream_ frontier as part of the pipeline, and we could recover the process frontier by using the reclock collection to invert the mz-timestamp domain data frontier. I don't think that latter part is possible in sql at right now (well, maybe its a non-`STRICT SERIALIZABLE` join between the data and reclock collections? 

Ill definition mention this in the doc